File: packages/shared/skill-defaults.js
const manifest = require("./skill-defaults.json");

function normalizeIds(input) {
  return Array.isArray(input)
    ? [...new Set(input.map((value) => String(value || "").trim()).filter(Boolean))]
    : [];
}

const SYSTEM_AUTO_SKILL_IDS = Object.freeze(normalizeIds(manifest.systemAutoSkillIds));
const GENERIC_ASSISTANT_SKILL_IDS = Object.freeze(normalizeIds(manifest.genericAssistantSkillIds));

function resolveEffectiveSkillIds(capabilities = {}, options = {}) {
  const value = capabilities && typeof capabilities === "object" ? capabilities : {};
  const disabled = new Set(normalizeIds(value.disabledSkills || value.disabled_skills));
  const ordered = [];
  const seen = new Set();
  const append = (ids, { ignoreDisabled = false } = {}) => {
    for (const id of normalizeIds(ids)) {
      if ((!ignoreDisabled && disabled.has(id)) || seen.has(id)) continue;
      seen.add(id);
      ordered.push(id);
    }
  };

  const inherits = value.inheritEngineDefaults !== false && value.inherit_engine_defaults !== false;
  if (inherits) append(SYSTEM_AUTO_SKILL_IDS);
  append(options.presetSkillIds || options.preset_skill_ids);
  append(value.enabledSkills || value.enabled_skills);
  append(options.selectedSkillIds || options.selected_skill_ids, { ignoreDisabled: true });
  return ordered;
}

module.exports = {
  version: Number(manifest.version || 0),
  SYSTEM_AUTO_SKILL_IDS,
  GENERIC_ASSISTANT_SKILL_IDS,
  resolveEffectiveSkillIds
};

File: crates/mia-core-common/src/skill_defaults.rs
use serde::Deserialize;
use std::collections::HashSet;
use std::sync::OnceLock;

const SKILL_DEFAULTS_JSON: &str = include_str!("../../../packages/shared/skill-defaults.json");

#[derive(Debug, Deserialize)]
#[serde(rename_all = "camelCase")]
struct SkillDefaultsManifest {
    version: u32,
    system_auto_skill_ids: Vec<String>,
    generic_assistant_skill_ids: Vec<String>,
}

fn manifest() -> &'static SkillDefaultsManifest {
    static MANIFEST: OnceLock<SkillDefaultsManifest> = OnceLock::new();
    MANIFEST.get_or_init(|| {
        let parsed: SkillDefaultsManifest = serde_json::from_str(SKILL_DEFAULTS_JSON)
            .expect("shared skill-defaults.json must be valid");
        assert_eq!(
            parsed.version, 1,
            "unsupported shared skill defaults version"
        );
        parsed
    })
}

pub fn system_auto_skill_ids() -> &'static [String] {
    &manifest().system_auto_skill_ids
}

pub fn generic_assistant_skill_ids() -> &'static [String] {
    &manifest().generic_assistant_skill_ids
}

pub fn resolve_effective_skill_ids(
    inherit_engine_defaults: bool,
    preset_skill_ids: &[String],
    enabled_skill_ids: &[String],
    disabled_skill_ids: &[String],
    selected_skill_ids: &[String],
) -> Vec<String> {
    let disabled = disabled_skill_ids
        .iter()
        .map(|id| id.trim())
        .filter(|id| !id.is_empty())
        .collect::<HashSet<_>>();
    let mut seen = HashSet::new();
    let mut resolved = Vec::new();

    let mut append = |ids: &[String], ignore_disabled: bool| {
        for id in ids {
            let id = id.trim();
            if id.is_empty()
                || (!ignore_disabled && disabled.contains(id))
                || !seen.insert(id.to_owned())
            {
                continue;
            }
            resolved.push(id.to_owned());
        }
    };

    if inherit_engine_defaults {
        append(system_auto_skill_ids(), false);
    }
    append(preset_skill_ids, false);
    append(enabled_skill_ids, false);
    append(selected_skill_ids, true);
    resolved
}
